### PR TITLE
7byiccoh: remove an existing service

### PR DESCRIPTION
**Pull request by jkunle**
**What**
This is one-half of [Enable editing removing of services](https://trello.com/c/yIYeQzAs/696-enable-editing-removing-of-services). This half addresses removing a service

**Why**
Users need to have the ability to remove a service

### CODE REVIEW

File: app/controllers/services_controller.rb [hunk 1]
```diff
@@ -25,6 +25,13 @@ def create
     end
   end
 
+  def destroy
+    service = Service.find_by_id(params[:id])
+    DeleteServiceEvent.create(service: service, data: { name: service.name, entity_id: service.entity_id })
+    flash[:notice] = "#{service.name} #{t('common.delete_successful')}"
```
Comment (jakubmiarka) [inline app/controllers/services_controller.rb]:
```suggestion
    flash[:notice] = t('common.delete_successful', name: service.name)
```

Comment (jakubmiarka) [reply app/controllers/services_controller.rb]: also, it probably should be `[:success]` and only if the event was valid

File: spec/models/delete_service_event_spec.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  let(:data) { { name: service.name, entity_id: service.entity_id } }
+  context 'on success' do
+    it 'event is persisted' do
+      delete_service_event = create(:delete_service_event, data: data)
+      expect(delete_service_event).to be_valid
+      expect(delete_service_event).to be_persisted
+    end
+  end
+  context 'on failure' do
+    it 'when name is not present' do
```
Comment (jakubmiarka) [inline spec/models/delete_service_event_spec.rb]: we're not passing name or entity_id to the event, so they don't need to be validated/expected

File: app/models/delete_service_event.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class DeleteServiceEvent < AggregatedEvent
+  belongs_to_aggregate :service
+
+  data_attributes :entity_id, :name
+  validates_presence_of :entity_id, :name
```
Comment (jakubmiarka) [inline app/models/delete_service_event.rb]:
```suggestion
```
we're only passing an existing service object to the event, no attributes or anything, so no need for it

File: spec/models/delete_service_event_spec.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+RSpec.describe DeleteServiceEvent, type: :model do
+  context 'on success' do
+    it 'is persisted' do
+      delete_service_event = create(:delete_service_event)
+      expect(delete_service_event).to be_valid
+      expect(delete_service_event).to be_persisted
+    end
+    it 'when event is called, service is deleted' do
+      existing_service = create(:service)
+      expect(Service.find_by_id(existing_service)).not_to be_nil
```
Comment (jakubmiarka) [inline spec/models/delete_service_event_spec.rb]:
```suggestion
      expect(Service.exists?(existing_service.id)).to be true
```
etc

File: spec/controllers/services_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      expect(response).to have_http_status(:success)
+    end
+  end
+
+  describe 'DELETED #destroy' do
+    it 'removes an existing service' do
+      service = create(:service, sp_component_id: sp_component)
+      expect {
+        delete :destroy, params: { id: service.id }
+      }.to change(Service, :count).by(-1)
```
Comment (jakubmiarka) [inline spec/controllers/services_controller_spec.rb]: this should be more robust, relying on count feels too fragile. We should test that the service we want to delete gets actually deleted...

File: spec/controllers/services_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+require 'rails_helper'
+require 'securerandom'
```
Comment (jakubmiarka) [inline spec/controllers/services_controller_spec.rb]: this is not used, is it?

File: app/controllers/services_controller.rb [hunk 7]
```diff
@@ -25,6 +25,17 @@ def create
     end
   end
 
+  def destroy
+    service = Service.find_by_id(params[:id])
+    return flash[:notice] = t('common.not_found', name: 'Service') unless service.present?
```
Comment (jakubmiarka) [inline app/controllers/services_controller.rb]: you can't just return here without rendering or redirecting

Comment (jakubmiarka) [inline app/controllers/services_controller.rb]: and it should be error, not notice

File: app/controllers/services_controller.rb [hunk 8]
```diff
@@ -25,6 +25,17 @@ def create
...
+
+    event = DeleteServiceEvent.create(service: service, data: { name: service.name, entity_id: service.entity_id })
+    return flash[:error] = event.errors.full_messages.join(', ') if event&.errors.present?
```
Comment (jakubmiarka) [inline app/controllers/services_controller.rb]: we don't need to this complicated, get inspired by the component deletion https://github.com/alphagov/verify-self-service/blob/ab9f4cd4f188f4ec31a4225d61373162364d1571/app/controllers/msa_components_controller.rb#L52

there's no reason for the event to be invalid as long as the service exists, so we should only need to check for the service existence and then fire the event.. all it does is deletion, no validation or anything...

Comment (jkunle) [reply app/controllers/services_controller.rb]: This was what I didn't on the first round must have misinterpreted you comment. And gone down this rabbit hole.  "and only if the event was valid" one of you first comments on friday

Comment (jakubmiarka) [reply app/controllers/services_controller.rb]: it was a general comment about validity... sorry, my bad, i didn't mean the literal object's validity...

File: app/controllers/services_controller.rb [hunk 9]
```diff
@@ -25,6 +25,15 @@ def create
     end
   end
 
+  def destroy
+    service = Service.find_by_id(params[:id])
+    if service.present?
+      DeleteServiceEvent.create(service: service, data: { name: service.name, entity_id: service.entity_id })
+      flash[:success] = t('common.action_successful', name: service.name, action: :deleted)
+    end
```
Comment (jakubmiarka) [inline app/controllers/services_controller.rb]: We should have the `else` service not present to be consistent with other destroy methods (i.e. component deletion), but it's a small thing.

File: app/views/admin/_service.erb [hunk 10]
```diff
@@ -27,7 +27,10 @@
         <td class="govuk-table__cell"><%= service.updated_at %></td>
       </tr>
       <tr class="govuk-table__row">
-        <td class="govuk-table__cell govuk-table__cell--numeric" colspan="2"><%= link_to 'Manage', '', class: 'govuk-button', data: { module: 'govuk-button' } %></td>
+        <td class="govuk-table__cell govuk-table__cell--numeric" colspan="2">
+        <%= link_to 'Manage', '', class: 'govuk-button', data: { module: 'govuk-button' } %>
+        <%= link_to 'Delete', service_path(service), method: :delete, data: { confirm: 'Are you sure?' }, class: 'govuk-button govuk-button--warning' %>
```
Comment (Rosa-Fox) [inline app/views/admin/_service.erb]: 'Delete' could be localised... also 'Manage' from above has slipped through from a previous PR

File: spec/controllers/services_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    end
+  end
+
+  describe 'DELETED #destroy' do
+    it 'expects DeleteServiceEvent to be called when deleting service' do
+      expect(DeleteServiceEvent).to receive(:create).and_call_original
+      delete :destroy, params: { id: service.id }
+      expect(Service.exists?(service.id)).to be false
+    end
+    it 'displays service that was removed' do
```
Comment (Rosa-Fox) [inline spec/controllers/services_controller_spec.rb]: maybe reword to 'displays service that was removed in flash message'... only because on first read I thought it would mean that it tests that the service appeared on the body of the page

File: app/views/admin/_service.erb (hunk 10 above)
Comment (jkunle) [reply app/views/admin/_service.erb]: Didn’t think it was important to local the admin view see jakub https://github.com/alphagov/verify-self-service/pull/130 that’s why i didn’t localise 254

File: spec/controllers/services_controller_spec.rb (hunk 11 above)
Comment (jkunle) [reply spec/controllers/services_controller_spec.rb]: Ok

Comment (jkunle) [reply spec/controllers/services_controller_spec.rb]: If you don't mind I'll revisit this, in a BAU which I have discussed with Jakub. This would cover else branch Jakub mentioned above.

File: app/views/admin/_service.erb (hunk 10 above)
Comment (Rosa-Fox) [reply app/views/admin/_service.erb]: 👍
